Write documentation/contributing guidelines for developers (#67)
Huawei-CPLLab/PyDSL: dev/dev_docs.md

The purpose of this document is to give you a high-level overview of the
internals of PyDSL, and some tips for developing the compiler.

> This document assumes you have already read some of the user facing

Reviewer: ```suggestion
This document assumes you have already read some of the user-facing
```

worry about storing them anywhere, since they are added to the context (which
is used to generate the output MLIR) automatically.

The visitor goes through the AST and converts every subtree to a PyDSL
representation.
It also has a scope stack and keeps track of what values are assigned to each
of the variables.
Generally, each variable defined by the user should correspond to some PyDSL
object (some type defined by PyDSL), never an MLIR binding object directly,
and rarely a normal built-in Python type.

## PyDSL types

PyDSL has many types whose purpose is to be a wrapper around an MLIR binding
type.
Examples of such types include `Int`, `Float`, `Index`, `Bool`, `MemRef`,
`Tensor`, `Tuple`, etc.

The way these types generally work is as follows.
They store information about the type of the object (e.g. bitwidth for `Int`,
shape and element_type for `Tensor`) and a variable called `value`, which is a
reference to the MLIR Python binding object.
This value will usually be something like "what operation in MLIR outputs this
value".

The `__init__` method of such a PyDSL type will usually expect an MLIR Python
binding object as its only argument.

> Casting between PyDSL types has... some support.
> `type.py` defines some things related to casting with `on_Call` and
> `Supportable`, which works okay I guess.

Reviewer: ```suggestion
Casting between PyDSL types has limited support.
`type.py` defines some things related to casting with `on_Call` and
`Supportable`, which works okay for now but might be improved in the future.
```
maybe something like this?
I think generally for documentation like this we should try to avoid using "I" and language this conversational. The language I think is less of an issue, but definitely the removal of "I" within documentation I think is important, since in theory other people will be maintaining these docs other than you in the future

    statement to run only a single test function.
- For generating arrays as test data, you should prefer to use `multi_arange`
from `helper.py`. You can do some operations on the output of this to make it
be in the correct range (e.g. see `test_linalg.py`).
    - Small arrays, like hardcoding a 2x2 is also okay.
    - You can also use other methods, like numpy random with a fixed seed if
    necessary.
    - I think `multi_arange` is slightly better than random, because it is
    easier to debug when something breaks, because when you print the
    arrays, it's easier to try to guess what went wrong when the input
    is highly structured like `multi_arange`.

# Formatting

You can run `hatch fmt -f --check` to check files for formatting issues and
`hatch fmt -f` to automatically fix them.
In the CI, we enforce that `hatch fmt -f --check` returns no unformatted files.

You can also run `hatch fmt --check` (no `-f`) to run the linter as well.
Note that this will show unsafe fixes as well, so you should not run it without
the `--check` flag, since it would break examples/testcases using docstring
directives.
We do not enforce that the linter passes in CI.

> I would generally recommend running `hatch fmt --check [filename]` on any files
> you are working on (after you are mostly done), and manually applying the fixes
> it suggests that you think are useful (I claim some of them are negative value
> and maybe we could add them to the ignore list).

Reviewer: ```suggestion
we recommend running `hatch fmt --check [filename]` on any files
you are working on (after you are mostly done), and manually applying the fixes
it suggests that you think are useful (some of them are potentially of negative value
and could be added to the ignore list in the future).
```

basically the same, which is nice.
    - The `on_Call` and `__call__` methods of `InlineFunction`s are designed
    specifically so that calling them from user code and compiler code is the
    same (no extra arguments need to be passed in either case).
    - Casting of scalar types works directly in both user and compiler code:
    `UInt32(UInt16(123))` is valid in both PyDSL and Python.
        - Maybe the behaviour from casting within the compiler will become
        a bit different depending on how
        https://github.com/Huawei-CPLLab/PyDSL/issues/19 is resolved.

# Circular import

Circular import errors can come up, I don't know a good way to fix them.
- You can try to restructure your function definitions so that no circular
imports are necessary (by moving the function to another file).
- If that is not possible, you can try importing inside a class, inside  a
function, or not at the top of the file (very ugly); or try `import xyz`
instead of `from xyz import abc`.

# Building a wheel

`hatch build` can be used to build a wheel and also a zip archive of the
project.
This can take a minute or two and there are not always proper "progress"
messages.

> I am not certain how well these work considering our dependency of MLIR
> Python bindings, which can not easily be packaged into a wheel as far as I
> understand.

Reviewer: ```suggestion
this may not work well considering our dependency of MLIR
Python bindings, which can not easily be packaged into a wheel.
```
